server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,428 @@
+                } else {
+                    logger.info("No legacy configurations of mirrors and credentials found in the project: {}.",
+                                project.name());
+                }
+            }
+        } catch (Exception ex) {
+            final MirrorMigrationException mirrorException = new MirrorMigrationException(
+                    "Failed to migrate mirrors and credentials. Rollback to the legacy configurations", ex);
+            try {
+                rollbackMigration();
+            } catch (Exception ex0) {
+                ex0.addSuppressed(mirrorException);
+                throw new MirrorMigrationException("Failed to rollback the mirror migration:", ex0);
+            }
+            throw mirrorException;
+        }
+
+        // Exit read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.WRITABLE))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Mirrors and credentials migration has been completed. (took: {} ms.)",
+                    stopwatch.elapsed().toMillis());
+
+        shortWords = null;
+    }
+
+    private boolean wasMigrated() throws Exception {
+        for (Project project : projectManager.list().values()) {
+            final MetaRepository repository = project.metaRepo();
+            if (getMetaData(repository, PATH_LEGACY_MIRRORS_BACKUP) != null) {
+                // The mirrors.json file has been backed up.
+                return true;
+            }
+            if (getMetaData(repository, PATH_LEGACY_CREDENTIALS_BACKUP) != null) {
+                // The credentials.json file has been backed up.
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean migrateMirrors(MetaRepository repository) throws Exception {
+        final ArrayNode mirrors = getMetaData(repository, PATH_LEGACY_MIRRORS);
+        if (mirrors == null) {
+            return false;
+        }
+
+        final List<MirrorCredential> credentials = repository.credentials()
+                                                             .get(30, TimeUnit.SECONDS);
+
+        final Set<String> mirrorIds = new HashSet<>();
+        for (JsonNode mirror : mirrors) {
+            if (!mirror.isObject()) {
+                logger.warn("A mirror config must be an object: {} (project: {})", mirror,
+                            repository.parent().name());
+                continue;
+            }
+            try {
+                migrateMirror(repository, (ObjectNode) mirror, mirrorIds, credentials);
+            } catch (Exception e) {
+                logger.warn("Failed to migrate a mirror config: {} (project: {})", mirror,
+                            repository.parent().name(), e);
+            }
+        }
+        // Back up the old mirrors.json file and don't use it anymore.
+        rename(repository, PATH_LEGACY_MIRRORS, PATH_LEGACY_MIRRORS_BACKUP, false);
```

Comment:
Proposal) I'm curious if the old `meta/[mirrors|credentials].json` will be visible to users in the new mirroring UI. If not, what do you think of doing the rename/backup migrations after a few versions to ensure stability?

I realize that changes to the new mirroring/credentials won't be reflected in the old mirrors (unless we pre-implement and release some logic). I still think it's better than nothing and provides us an escape hatch if something goes wrong in the new release (e.g. it may not necessarily be due to this feature, but an armeria bug may force us to revert versions)

Reply:
Hmm. It's a good idea, but it's not easy to apply. What if a bug is discovered a certain amount of time after the new mirroring configuration is exposed to users?

If an update has already been made in the new file, converting to the old file does not seem to be a good option. The migration itself should be verified through testing, and for bugs caused by external libraries, it would be better to revert the version only rather than reverting the all changes.

Speaking of the release strategy, what do you think of releasing a new version only for this change to minimize side effects?

Reply:
> Hmm. It's a good idea, but it's not easy to apply. What if a bug is discovered a certain amount of time after the new mirroring configuration is exposed to users?

I see. I was thinking there is a higher probability that a bug is discovered right away rather than a long time. This comment also assumed that mirroring configurations are usually not updated often.
 
I understood we prefer the current approach - if a roll back to the previous version is needed, no mirrors will be executed.